package gsbn;

message SolverParam{
	required GenParam gen_param=1;
	required NetParam net_param=3;
}

message GenParam{
	optional string stim_file=1;
	repeated ModeParam mode_param=2;
}

message ModeParam{
	required uint32 begin_time=2;
	required uint32 end_time=3;
	required uint32 type=1;
	optional uint32 stim_index=4;
}

message NetParam{
	repeated PopParam pop_param=1;
	repeated ProjParam proj_param=2;
	required ConfParam conf_param=3;
}

message PopParam{
	required uint32 pop_num=1;
	repeated HcuParam hcu_param=2;
}

message HcuParam{
	required uint32 hcu_num=1;
	required uint32 slot_num=2;
	repeated McuParam mcu_param=3;
}

message McuParam{
	required uint32 mcu_num=1;
	required uint32 fanout_num=2;
}
		
message ProjParam{
	required uint32 src_pop = 1;
	required uint32 dest_pop = 2;
}

message ConfParam{
	required float kp = 1;
	required float ke = 2;
	required float kzj = 3;
	required float kzi = 4;
	required float kftj = 5;
	required float kfti = 6;
	required float bgain = 7;
	required float wgain = 8;
	required float wtagain = 9;
	required float igain = 10;
	required float eps = 11;
	required float lgbias = 12;
	required float snoise = 13;
	required float maxfqdt = 14;
	required float taumdt = 15;
}

message SolverState{
	required uint32 timestamp=1;
	repeated TableState table_state=2;
}

message TableState{
	required string name=1;
	required bytes desc=2;
	required bytes data=3;
}

message StimRawData{
	repeated float data=1 [packed=true];
}
